### PR TITLE
9560 eliminate loop for surface heat transfer algorithm and replace with select reset

### LINKED ISSUE #9560: Test for surface heat balance algorithm should be moved to some other place?
Issue overview
--------------
In ZoneTempPredictorCorrector::CalcPredictedHumidityRatio this loop is executed every time step. If this is in fact the same result each time (i.e., the surface heat balance algorithm cannot change during the simulation) then this slows program exectution. This could be moved to an init routine where an array variable is set as no_ht_EMPD_or_HAMT(ZoneNum).

UPDATE: further investigation shows that the SumHmA* variables are only used for EMPD and HAMT so they can be reset where these values are calculated.

        // if no surface in the zone uses EMPD or HAMT then zero
        bool no_ht_EMPD_or_HAMT(true);
        for (int i = Zone(ZoneNum).HTSurfaceFirst, e = Zone(ZoneNum).HTSurfaceLast; i <= e; ++i) {
            auto const &htAlgo(state.dataSurface->Surface(i).HeatTransferAlgorithm);
            if ((htAlgo == DataSurfaces::HeatTransferModel::EMPD) || (htAlgo == DataSurfaces::HeatTransferModel::HAMT)) {
                no_ht_EMPD_or_HAMT = false;
                break;
            }
        }
        if (no_ht_EMPD_or_HAMT) {
            state.dataHeatBalFanSys->SumHmARaW(ZoneNum) = 0.0;
            state.dataHeatBalFanSys->SumHmARa(ZoneNum) = 0.0;
        }

HeatBalanceAlgorithm,
       \memo Determines which Heat Balance Algorithm will be used ie.
       \memo CTF (Conduction Transfer Functions),
       \memo EMPD (Effective Moisture Penetration Depth with Conduction Transfer Functions).
       \memo Advanced/Research Usage: CondFD (Conduction Finite Difference)
       \memo Advanced/Research Usage: ConductionFiniteDifferenceSimplified
       \memo Advanced/Research Usage: HAMT (Combined Heat And Moisture Finite Element)
       \unique-object
       \format singleLine
  A1 , \field Algorithm
       \type choice
       \key ConductionTransferFunction
       \key MoisturePenetrationDepthConductionTransferFunction
       \key ConductionFiniteDifference
       \key CombinedHeatAndMoistureFiniteElement
       \default ConductionTransferFunction

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version)
 - Version of EnergyPlus (if using an intermediate build, include SHA)
 - Unmethours link or helpdesk ticket number

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (example file FurnaceWithDXSystemRHcontrol.idf should hit these lines of code)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by rraustad**
Pull request overview
---------------------
 - Fixes #9560
 - Legacy code has been removed and replace by appropriate reset where EMDP and HAMT SumHmA* variables are calculated.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] ~If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix~
 - [ ] ~If feature, test running new feature, try creative ways to break it~
 - [x] CI status: all green or justified
 - [x] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [x] Check any new function arguments for performance impacts
 - [ ] ~Verify IDF naming conventions and styles, memos and notes and defaults~
 - [ ] ~If new idf included, locally check the err file and other outputs~

### CODE REVIEW

File: src/EnergyPlus/ZoneTempPredictorCorrector.cc
Comment (amirroth) [inline src/EnergyPlus/ZoneTempPredictorCorrector.cc:4565]: This may be more cleanup than you want to do in this PR, but at some point you may want to transition all of the zone-level variables in DataHeatBalFanSys from 76 (not a made-up number) arrays of scalars to 1-5 arrays of `struct` (grouped by variable affinity). I know that I may have previously implied that arrays of scalars are superior to arrays of `struct` but that is context/usage-pattern specific and also depends on the size of the struct. In this specific case, I think array management overhead is dominant.

**Comment by rraustad** [comment]
> Two things about this function. First, in x86_64 only the first six parameters are passed through registers

This is turning into a hugh rabbit hole. I have mostly converted over to arrays but I think it better to merge this no diff change after I correct the warnings.

**Comment by mjwitte** [comment]
Yeah, this is likely to have conflicts with #9553 and #9394. Let's minimize sweeping changes in predictor corrector until some of those are merged or try to sequence things somehow.

**Comment by amirroth** [comment]
Rome was not built in a single PR.

File: src/EnergyPlus/DataHeatBalance.hh
Comment (mjwitte) [inline src/EnergyPlus/DataHeatBalance.hh]: Instead of a new `Array1D` why not just add this as a field in Zone? And maybe name it `noEMPDorHAMT`?

**Review by mjwitte — COMMENTED** [review]
I've merged in develop, resolved and repaired conflicts, and moved the `spacePowerReductionFactor.dimension`s to `AllocateIntGains`. Unit tests all pass locally.

Let's wait for a clean CI report before any other changes.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (mjwitte) [inline src/EnergyPlus/HeatBalanceSurfaceManager.cc]: This can be wrapped in `if (state.dataHeatBal->AnyEMPD || state.dataHeatBal->AnyHAMT)` No need to search the surfaces if there isn't any EMPD or HAMT in the entire simulation.

File: src/EnergyPlus/DataHeatBalance.hh
Comment (mjwitte) [inline src/EnergyPlus/DataHeatBalance.hh]: This should be initialized to `true`. So why didn't this cause diffs? Do those calcs even get used for anything?

Comment (rraustad) [reply src/EnergyPlus/DataHeatBalance.hh]: I saw that and fixed it. I am confused at why this did not cause more problems. The initial Array1D\<bool\>no_ht_EMPD_or_HAMT was never initialized and just happened to be true. This is used to reset humidity variables in 2 places in ZoneTempPredictorCorrector. Only the unit test identified a problem with my implementation, but why didn't this cause more diffs is a good question.

        if (no_ht_EMPD_or_HAMT) {
            state.dataHeatBalFanSys->SumHmARaW(ZoneNum) = 0.0;
            state.dataHeatBalFanSys->SumHmARa(ZoneNum) = 0.0;
        }

File: src/EnergyPlus/ZoneTempPredictorCorrector.cc
Comment (mjwitte) [inline src/EnergyPlus/ZoneTempPredictorCorrector.cc]: Searching the code for SumHmARaW, I don't think these resets should be here, or at line 5637. I think these are left over from way back before EMPD and HAMT were added.  I'm going to check the archive to be sure. There's another place [here in `CalcHeatBalanceAir`](https://github.com/NREL/EnergyPlus/blob/fb5dce86f0a63aa46f11c9ddacdd944e5f247bd5/src/EnergyPlus/HeatBalanceAirManager.cc#L4782-L4785) that's also resetting these to zero with no check for EMPD or HAMT.  If these are appropriately place, then maybe we ditch the zone level flags and put if(AnyEMPD || AnyHAMT) around this block and dump the zone-level flags. And rewrite these bulk array assignments as loops.

Comment (mjwitte) [reply src/EnergyPlus/ZoneTempPredictorCorrector.cc]: And the reset to zero in `CalcHeatBalanceAir` shouldn't be there either. It's from way back when and there used to be resets of SumHATSurf  = 0.0 there as well.
The only place now that `SumHmARa` and `SumHmARaW` are accumulated is [here in `CalcHeatBalanceInsideSurf2`](https://github.com/NREL/EnergyPlus/blob/fb5dce86f0a63aa46f11c9ddacdd944e5f247bd5/src/EnergyPlus/HeatBalanceSurfaceManager.cc#L7600-L7601). I would delete the zone-level `noEMPTorHAMT` flage and just reset all three of these before the surface loop:
```
        state.dataHeatBalFanSys->SumHmAW(zoneNum) = 0.0;
        state.dataHeatBalFanSys->SumHmARa(zoneNum) = 0.0;
        state.dataHeatBalFanSys->SumHmARaW(zoneNum) = 0.0;
```

Comment (rraustad) [reply src/EnergyPlus/ZoneTempPredictorCorrector.cc]: OK, I have 0 diffs on HAMT*.idfs and a few other files.

**Review by mjwitte — APPROVED** [review]
Soooooo, mannnny, greaaaaat, cleanups here! CI is all green for go. Another 0.13% performance improvement. Merging.

**Pull request merged**
